csfieldguide/static/interactives/algorithm-timer/js/algorithm-timer.js

```diff
@@ -0,0 +1,177 @@
+const Mathjs = require('mathjs');
+
+const COMPLEXITY_TEXT = {
+  'log': 'log(n)',
+  'nlog': 'nlog(n)',
+  'linear': 'n',
+  'squared': 'n&sup2;',
+  'cubed': 'n&sup3;',
+  'fourth-power': 'n<sup>4</sup>',
+  'sixth-power': 'n<sup>6</sup>',
+  'exponential': '2<sup>n</sup>',
+  'factorial': 'n!',
+};
+
+const TIME_SCALERS = {
+  'nanoseconds': 1000000000,
+  'microseconds': 1000000,
+  'milliseconds': 1000,
+  'seconds': 1,
+  'minutes': Mathjs.divide(1, 60),
+  'hours': Mathjs.divide(1, 3600),
+  'days': Mathjs.divide(1, 86400),
+  'years': Mathjs.divide(1, 31557600),
+  'centuries': Mathjs.divide(1, 3155760000),
+};
+
+
+$(document).ready(function() {
+  var complexity = $('input[name=complexity]:checked').prop('id');
+  var resultForm = $('input[name=result-form]:checked').prop('id');
+  var n = $('#n-items').val();
+  var speed = $('#speed').val();
+  var processors = $('#processors').val();
+  var timeUnits = $('input[name=time]:checked').prop('id');
+  updateData()
+
+  $('input[name=complexity]').click(function() {
+    complexity = $('input[name=complexity]:checked').prop('id');
+    chosenComplexityText = COMPLEXITY_TEXT[complexity];
+    $('#complexity-chosen').html(chosenComplexityText);
+    updateData()
+  });
+  $('input[name=result-form]').click(function() {
+    resultForm = $('input[name=result-form]:checked').prop('id');
+    updateData()
+  });
+  $('#n-items').on('input', function() {
+    n = $('#n-items').val();
+    if (inputIsValid(n, speed, processors)) {
+      updateData()
+    }
+  });
+  $('#speed').on('input', function() {
+    speed = $('#speed').val();
+    if (inputIsValid(n, speed, processors)) {
+      updateData()
+    }
+  });
+  $('#processors').on('input', function() {
+    processors = $('#processors').val();
+    if (inputIsValid(n, speed, processors)) {
+      updateData()
+    }
+  });
+  $('input[name=time]').click(function() {
+    timeUnits = $('input[name=time]:checked').prop('id');
+    updateData()
+  });
+
+  /** Get the time it will take for the algorithm to finish executing
+   *  and display it. */
+  function updateData() {
+    var result = calculateTimeTaken(complexity, resultForm, n, speed, processors, timeUnits);
+    $('#output').val(result + ' ' + timeUnits);
+  }
+});
+
+
+/** Calculates the time the algorithm will take to finish
+ *  executing in the selected time units */
+function calculateTimeTaken(complexity, resultForm, n, speed, processors, timeUnits) {
+  // can only have whole integers for number of items
+  n = Mathjs.bignumber(Math.round(n));
+  $('#n-items').val(n);
+  speed = Mathjs.bignumber(speed);
+  processors = Mathjs.bignumber(processors);
+  var steps;
+  
+  if (complexity == 'log') {
+    steps = Mathjs.bignumber(Mathjs.log(n));
+  } else if (complexity == 'linear') {
+    steps = Mathjs.bignumber(n);
+  } else if (complexity == 'nlog') {
+    steps = Mathjs.bignumber(n * Mathjs.log(n));
```

Comment:
Yeah regarding earlier, these log(n) numbers should be `1 + number-rounded-up`
1 + log(1) == 1 operation
1 + log(2) == 2 operations
1 + log(3) == 2 operations
1 + log(4) == 2 operations
etc

1 + 1(log(1)) == 1 operation
1 + 2(log(2)) == 2 operations
1 + 3(log(3)) == 3 operations
etc